test/e2e/specs/site-editor/font-library.spec.js

```diff
@@ -15,15 +15,66 @@ test.describe( 'Font Library', () => {
 		} );
 
 		test( 'should display the "Manage Fonts" icon', async ( { page } ) => {
-			await page.getByRole( 'button', { name: /styles/i } ).click();
+			await page.getByRole( 'button', { name: 'Styles' } ).click();
 			await page
-				.getByRole( 'button', { name: /typography styles/i } )
+				.getByRole( 'button', { name: 'Typography Styles' } )
 				.click();
 			const manageFontsIcon = page.getByRole( 'button', {
-				name: /manage fonts/i,
+				name: 'Manage Fonts',
 			} );
 			await expect( manageFontsIcon ).toBeVisible();
 		} );
+
+		test( 'should allow user to upload a local font', async ( {
+			page,
+		} ) => {
+			await page.getByRole( 'button', { name: 'Styles' } ).click();
+			await page
+				.getByRole( 'button', { name: 'Typography Styles' } )
+				.click();
+			await page
+				.getByRole( 'button', {
+					name: 'Manage Fonts',
+				} )
+				.click();
+
+			// Delete test font family (Exo 2) if it exists
+			await expect(
+				page.getByRole( 'tab', { name: 'Upload' } )
+			).toBeVisible( { timeout: 80000 } );
+
+			if (
+				await page.getByRole( 'button', { name: 'Exo 2' } ).isVisible()
+			) {
+				await page.getByRole( 'button', { name: 'Exo 2' } ).click();
+				await page.getByRole( 'button', { name: 'Delete' } ).click();
+				await page.getByRole( 'button', { name: 'Delete' } ).click();
+				await page.getByRole( 'tab', { name: 'Library' } ).click();
+			}
+
+			// Upload local fonts
+			await page.getByRole( 'tab', { name: 'Upload' } ).click();
+			const fileChooserPromise = page.waitForEvent( 'filechooser' );
+			await page.getByRole( 'button', { name: 'Upload Font' } ).click();
+			const fileChooser = await fileChooserPromise;
+			// Provides coverage for https://github.com/WordPress/gutenberg/issues/59023.
+			await fileChooser.setFiles( [
+				'./test/e2e/assets/Exo2-Regular.woff',
+				'./test/e2e/assets/Exo2-SemiBoldItalic.woff2',
+			] );
+
+			// Check fonts were installed
+			await expect(
+				page
+					.getByLabel( 'Upload' )
+					.getByText( 'Fonts were installed successfully.' )
+			).toBeVisible( { timeout: 40000 } );
+			await page.getByRole( 'tab', { name: 'Library' } ).click();
+			// Provides coverage for https://github.com/WordPress/gutenberg/issues/60040.
```

Comment:
I wonder are we able to check that the necessary CSS rules were generated 🤔 ? Because that was the cause of the bug.

Reply:
Possible method.

- apply the given font to a element
- getByRole() an element with the font face applied
- get the computed value
```
  await someElement.evaluate((element) =>
    window.getComputedStyle(element).getPropertyValue('font-family'),
  );
```
- assert it has "Quotes"

Reply:
This does feel like we're straying outside of e2e test territory.

I think we may need to accept that this would be better handled by:

- adding PHP unit test for REST API which asserts fonts with numbers and spaces are correctly formatted with quotes.
- adding unit tests to the JS code which requests the font family declarations to ensure it's calling the endpoint in question.

I don't see that the above is achievable or indeed wise this late in the release. 

As a result, we may need to accept the current e2e test is the best level coverage we can achieve right now.

Reply:
Added a test for this here: https://github.com/WordPress/gutenberg/pull/60221/commits/5677dbe10b824c6061f05e03c7c16cdd751e56c9

Reply:
Noting that we have both [PHP](https://github.com/WordPress/wordpress-develop/blob/9bf06950a115584c498adba125eb3c233f878775/tests/phpunit/tests/fonts/font-library/wpFontUtils/sanitizeFontFamily.php) and [JS](https://github.com/WordPress/gutenberg/blob/6f9b2903e7a8592b7c77977110301524ea35a9b6/packages/edit-site/src/components/global-styles/font-library-modal/utils/test/preview-styles.spec.js#L127) unit tests for font family formatting. I think the original problem was that we we using the font-family values parsed from the uploaded font files, rather than what was returned from the API. (@matiasbenedetto , please correct me if I'm wrong).